continuationsPlugin/src/main/scala/continuations/DefDefTransforms.scala

```diff
@@ -60,47 +60,88 @@ object DefDefTransforms extends TreesChecks:
 
   private def transformSuspendContinuationBody(
       callSuspensionPoint: tpd.Tree,
-      safeContinuationRef: tpd.Tree)(using Context): List[tpd.Tree] = {
+      safeContinuationRef: tpd.Tree)(using Context): tpd.Tree = {
     val resumeMethod = safeContinuationRef.select(termName("resume"))
 
-    val suspendContinuationBody: List[tpd.Tree] =
-      callSuspensionPoint
-        .filterSubTrees {
-          case Trees.DefDef(_, _, _, _) => true
-          case _ => false
-        }
-        .flatMap {
-          case Trees.DefDef(_, _, _, Trees.Block(stats, expr)) =>
-            stats.map(s => s.withType(s.typeOpt)) :+ expr.withType(expr.typeOpt)
-          case tree =>
-            tree.asInstanceOf[tpd.DefDef].rhs :: Nil
-        }
-
     def keepSubTree(t: tpd.Tree): Boolean =
       t.symbol.exists &&
         t.symbol.owner.isAnonymousFunction &&
         t.symbol.owner.paramSymss.exists(_.exists(hasContinuationClass))
 
-    val removedAnonFunc: List[Symbol] = suspendContinuationBody
-      .flatMap(_.filterSubTrees(keepSubTree))
-      .map(_.symbol.owner)
-      .distinct
+    val anonymousContinuationFunctions = callSuspensionPoint.filterSubTrees {
+      case t @ Trees.DefDef(_, _, _, _) if keepSubTree(t)=> true
+      case tree @ Trees.Select(qualifier, name)
+          if tree.symbol.owner.name.show == "Continuation" && name.show == "resume" =>
+        true
+      case t =>
+        false
+    }
 
-    val suspendContinuationResume = new TreeTypeMap(
+    val continuationReferences = anonymousContinuationFunctions.map(_.symbol.owner)
+
+    val continuationReferencesSize = continuationReferences.size
+    val safeContinuationRefSymbols =
+      List.fill(continuationReferences.size)(safeContinuationRef.symbol)
+    val safeContinuationRefOwners = safeContinuationRefSymbols.map(_.owner)
+
+    val hypothesis = new TreeTypeMap(
       treeMap = {
-        case tree @ Trees.Apply(_, List(resumeArg)) if treeCallsResume(tree) =>
-          resumeMethod.appliedTo(resumeArg)
-        case tree =>
-          tree
+        case t @ Trees.DefDef(_, _, _, _)
+            if t.symbol.isAnonymousFunction && t
+              .symbol
+              .paramSymss
+              .flatten
+              .exists(_.info.hasClassSymbol(requiredClass(continuationFullName))) =>
```

Comment:
We can use the function that was refactored above: 
```suggestion
        case t @ Trees.DefDef(_, _, _, _) if keepSubTree(t) =>
```

BTW if you know a more informative name for that auxiliary predicate than `keepSubTree`, feel free to change it.:)